# Importing Libraries and files

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = "X:\Projects\Self-Study\Career Foundry\Data Immersion\Achievement 4"

In [3]:
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'))

In [4]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','orders_wrangled.csv'), index_col=False)

In [5]:
df_ords.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06       3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01       2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01       2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00       1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01       3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02       6.000000e+00   

       order_hour_of_day  days since_last_order  
count       3.421083e+06           3.214874e+06  
mean        1.345202e+01           1.111484e+01  
std         4.226088e+00           9.206737e+00  
min         0.000000e+00           0.000000e+00  
25%         1.000000e+01           4.000000e+00  
50%         1.300000e+01           7.000000e+00  
75%         1.600000e+01           1.500000e+01  
max         2.300000e+01           3.000000e+01

In [6]:
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

# Creating new test df for fixing mixed-type columns

In [ ]:
# Create a dataframe
df_test = pd.DataFrame()

In [ ]:
# Create mixed-type column
df_test['mix']=['a','b', 1, True]

In [ ]:
df_test

In [ ]:
# check for mixed-types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [ ]:
df_test.dtypes

In [ ]:
df_test['mix'] = df_test['mix'].astype('str')

In [ ]:
df_test.dtypes

# Missing values

In [ ]:
# Finding missing values
df_prods.isnull().sum()

In [ ]:
# Creating subset to view missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [ ]:
df_nan

In [ ]:
# Testing drop rows that contain NaN. View only
df_nan.dropna(inplace = False)

In [ ]:
# Testing drop NaN from col. View only
df_nan.dropna(subset = ['product_name'], inplace = False)

### Address Missing Value 1

In [7]:
df_prods.shape

(49693, 5)

In [8]:
# creating new df
df_prods_clean = pd.DataFrame()

In [9]:
# create subset with non-missing data (remove NaN rows)
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [10]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [11]:
df_prods_clean.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Length: 49677, dtype: bool

In [12]:
# Creates subset of duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [13]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### Addressing Duplicates

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# dropping duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting

In [17]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked.csv'), index=False)

# Step 2 for df_prods

In [18]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [19]:
# checking data using describe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

Maximum price for a product is 99,999 which is suspicious for grocery item

# Step 3 mixed-type data for df_prods

In [20]:
df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

In [21]:
# checking for mixed-type data

for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

product_name


# Step 4

Product_name is identified as an object data type, which is consistent with the 'str' that we want the type of data to be in

# Step 5 missing values for df_ords

In [22]:
# Checking df_ords
df_ords.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days since_last_order  
0                    NaN  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [23]:
# Finding missing values for df_ords
df_ords.isnull().sum()

order_id                      0
user_id                       0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days since_last_order    206209
dtype: int64

There are 20,6209 cases of missing values in day_since_last_order. This value could be that a customer just recently bought less than a day ago, or a new customer that has never bought before.

In [24]:
# Creating subset to view missing values
df_ords_nan = df_ords[df_ords['days since_last_order'].isnull()==True]

In [25]:
df_ords_nan

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
11        2168274        2             1                  2   
26        1374495        3             1                  1   
39        3343014        4             1                  6   
45        2717275        5             1                  3   
...           ...      ...           ...                ...   
3420930    969311   206205             1                  4   
3420934   3189322   206206             1                  3   
3421002   2166133   206207             1                  6   
3421019   2227043   206208             1                  1   
3421069   3154581   206209             1                  3   

         order_hour_of_day  days since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 6 columns]

# Step 6 address missing values

It is proposed to create a flag to identify the customers with missing days since_last_order value

In [26]:
# Creating a Flag for customers with NaN value
df_ords_NanFlag = df_ords
df_ords_NanFlag['NaN_Flag'] = df_ords['days since_last_order'].isnull()==True

In [27]:
df_ords_NanFlag

order_id  user_id  order_number  order_day_of_week  \
0         2539329        1             1                  2   
1         2398795        1             2                  3   
2          473747        1             3                  3   
3         2254736        1             4                  4   
4          431534        1             5                  4   
...           ...      ...           ...                ...   
3421078   2266710   206209            10                  5   
3421079   1854736   206209            11                  4   
3421080    626363   206209            12                  1   
3421081   2977660   206209            13                  1   
3421082    272231   206209            14                  6   

         order_hour_of_day  days since_last_order  NaN_Flag  
0                        8                    NaN      True  
1                        7                   15.0     False  
2                       12                   21.0     False  
3                        7                   29.0     False  
4                       15                   28.0     False  
...                    ...                    ...       ...  
3421078                 18                   29.0     False  
3421079                 10                   30.0     False  
3421080                 12                   18.0     False  
3421081                 12                    7.0     False  
3421082                 14                   30.0     False  

[3421083 rows x 7 columns]

# Step 7 & 8 Duplicates for df order

In [28]:
# creating subsets for duplicate
df_ords_NanFlag_dup = df_ords_NanFlag[df_ords_NanFlag.duplicated()]

In [29]:
df_ords_NanFlag_dup

Empty DataFrame
Columns: [order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days since_last_order, NaN_Flag]
Index: []

No duplicates found

# Export

In [30]:
df_ords_NanFlag.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_checked.csv' ), index=False)